File: Algorithm/Code_Materials/c_or_cpp/charpter8_tree_hot_problems/level3/lowest_common_ancestor.cpp
#include <vector>
#include <stack>
#include <iostream>
#include <queue>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode *lowestCommonAncestor(TreeNode *root, TreeNode *p, TreeNode *q)
{
    if (root == NULL || root == p || root == q)
    {
        return root;
    }
    TreeNode *left = lowestCommonAncestor(root->left, p, q);
    TreeNode *right = lowestCommonAncestor(root->right, p, q);
    if (left == NULL && right == NULL)
    {
        return NULL; // 1.
    }
    if (left == NULL)
    {
        return right; // 3.
    }
    if (right == NULL)
    {
        return left; // 4.
    }
    return root; // 2. if(left != null and right != null)
}

int main()
{
    // 创建测试用例的二叉树
    // 创建树根节点
    TreeNode *root = new TreeNode(1);

    // 创建节点p和q
    TreeNode *p = new TreeNode(3);
    TreeNode *q = new TreeNode(2);
    TreeNode *p1 = new TreeNode(5);
    TreeNode *p2 = new TreeNode(4);
    TreeNode *q1 = new TreeNode(6);

    // 创建节点p和q的子节点
    p->left = p1;
    p->right = p2;
    q->left = q1;

    // 将节点p和q插入到树中
    root->left = p;
    root->right = q;

   TreeNode *tree =lowestCommonAncestor(root, p, q)  ;
    return 0;
}

File: Algorithm/Code_Materials/Java/src/main/java/com/yugutou/charpter8_tree_hot_problems/level3/LowestCommonAncestor.java
package com.yugutou.charpter8_tree_hot_problems.level3;

import com.yugutou.tools.BinaryTree;
import com.yugutou.tools.TreeNode;

/**
 * LeetCode235.给定一个二叉搜索树,
 * 找到该树中两个指定节点的最近公共祖先。
 */
public class LowestCommonAncestor {
    public static void main(String[] args) {
        BinaryTree binaryTree = new BinaryTree();
        binaryTree.root = binaryTree.buildLowestCommonAncestor();

        TreeNode p = new TreeNode(0);
        TreeNode q = new TreeNode(5);
        TreeNode node = lowestCommonAncestor(binaryTree.root, p, q);
        System.out.println(node.val);
    }


    /**
     * 递归实现
     */
    public static TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        //如果小于等于0，说明p和q位于root的两侧，直接返回即可
        if ((root.val - p.val) * (root.val - q.val) <= 0)
            return root;
        //否则，p和q位于root的同一侧，就继续往下找
        return lowestCommonAncestor(p.val < root.val ? root.left : root.right, p, q);
    }

    /**
     * 迭代实现
     *
     * @param root
     * @param p
     * @param q
     * @return
     */
    public static TreeNode lowestCommonAncestor2(TreeNode root, TreeNode p, TreeNode q) {
        //如果根节点和p,q的差相乘是正数，说明这两个差值要么都是正数要么都是负数，也就是说
        //他们肯定都位于根节点的同一侧，就继续往下找
        while ((root.val - p.val) * (root.val - q.val) > 0)
            root = p.val < root.val ? root.left : root.right;
        //如果相乘的结果是负数，说明p和q位于根节点的两侧，如果等于0，说明至少有一个就是根节点
        return root;
    }
}
